File: purecoincidence/src/main/scala/purecoincidence/NoValueThing.scala
package purecoincidence

object NoValueThing {
  def divideA(x: Int, y: Int): Int = x / y

  divideA(8, 2)
  divideA(6, 3)
  //  divideA(3, 0) // => ArithmeticException

  def divideB(x: Int, y: Int): Int = {
    assert(y != 0)
    x / y
  }

  def divideC(x: Int, y: Int): Int = {
    if (y == 0) -1 // or 'null'
    else x / y
  }

  def divideD(x: Int, y: Int): (Int, Boolean) =
    if (y == 0) (0, false)
    else (x / y, true)

  def divide(x: Int, y: Int): Option[Int] =
    if (y == 0) None
    else Some(x / y)
}

File: hopefullypure/app/src/main/java/hopefullypure/NoValueThing.java
package hopefullypure;

import java.util.Optional;

public class NoValueThing {
    static int divideA(int x, int y) {
        return x / y;
    }

    static int divideB(int x, int y) {
        assert y != 0;
        return x / y;
    }

    static int divideC(int x, int y) {
        if (y == 0) return -1;
        return x / y;
    }

    static Optional<Integer> divide(int x, int y) {
        if (y == 0) return Optional.empty();
        else return Optional.of(x / y);
    }
}
